### PR TITLE
New rule: Color contrast enhanced

### PR DESCRIPTION
**Pull request by Jym77**
Rule for 1.4.6 Color Contrast (nehanced) built upon the existing rule for 1.4.3.
I hope I didn't mess up too much the contrast computation 😅 

Closes issue(s):

- N/A

Need for Final Call:
This will require a 2 weeks Final Call (new rule)

---

### **When creating PR:**

- [X] Make sure you're requesting to **pull a branch** (right side) to the `develop` branch (left side).
- [X] Make sure you do not remove the "How to Review and Approve" section in your pull request description

### **After creating PR:**

- [X] Add yourself (and co-authors) as "Assignees" for PR.
- [X] Add label to indicate if it's a `Rule`, `Definition` or `Chore`.
- [ ] [Link the PR to any issue it solves](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue). This will be done automatically by [referencing the issue](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) at the top of this comment in the indicated place.
- [ ] Optionally request feedback from anyone in particular by assigning them as "Reviewers".

### **When merging a PR:**

- [ ] Close any issue that the PR resolves. This will happen automatically upon merging if the PR was correctly linked to the issue, e.g. by referencing the issue at the top of this comment.

## How to Review And Approve

- Go to the “Files changed” tab
- Here you will have the option to leave comments on different lines.
- Once the review is completed, find the “Review changes” button in the top right, select “Approve” (if you are really confident in the rule) or "Request changes" and click “Submit review”.
- Make sure to also review the proposed Final Call period. In case of disagreement, the longer period wins.

### CODE REVIEW

File: _rules/text-contrast-enhanced-09o5cg.md [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+  authors:
+    - Jean-Yves Moyen
+---
+
+## Applicability
+
+This rule applies to any [visible][] character in a [text node][] that is a [child][] in the [flat tree][] of an HTML element, except if the [text node][] has an [ancestor][] in the [flat tree][] for which one of the following is true:
+
+- **widget**: the ancestor has a [semantic role][] that inherits from `widget`; or
+- **disabled label**: the ancestor is used in the [accessible name][] of a `widget` that is [disabled][]; or
```
Comment (carlosapaduarte) [inline _rules/text-contrast-enhanced-09o5cg.md:38]:
```suggestion
- **disabled label**: the ancestor is used in the [accessible name][] of a **widget** that is [disabled][]; or
```

I believe the "widget" in this sentence is referring to the widget condition that was expressed in the previous sentence. If it is referring to the "widget" semantic role then it needs to be rephrased because semantic roles do not have accessible names (elements do).

File: _rules/text-contrast-enhanced-09o5cg.md [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+- Implementation of [Presentational Roles Conflict Resolution][] varies from one browser or assistive technology to another. Depending on this, some elements can have a [semantic role][] of `none` and fail this rule with some technology but users of other technologies would not experience any accessibility issue.
+
+## Background
+
+Passing this rule does not mean that the text has sufficient color contrast. If all background pixels have a low contrast with all foreground pixels, the success criterion is guaranteed to not be satisfied. When some pixels have sufficient contrast, and others do not, legibility should be considered. There is no clear method for determining legibility, which is why this is out of scope for this rule.
+
+When the text color or background color is not specified in the web page, colors from other [origins][] will be used. Testers must ensure colors are not affected by styles from a [user origin][], such as a custom style sheet. Contrast issues cause by specifying the text color but not the background or vise versa, must be tested separately from this rule.
```
Comment (carlosapaduarte) [inline _rules/text-contrast-enhanced-09o5cg.md:64]:
```suggestion
When the text color or background color is not specified in the web page, colors from other [origins][] will be used. Testers must ensure colors are not affected by styles from a [user origin][], such as a custom style sheet. Contrast issues caused by specifying the text color but not the background or vice versa, must be tested separately from this rule.
```

File: _rules/text-contrast-enhanced-09o5cg.md [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+
+```html
+<p style="color: #000; font-size:18pt; background: #777;">
+	Some text in a human language
+</p>
+```
+
+#### Passed Example 5
+
+This 14pt bold black text has a contrast ratio of 3.6:1 on the gray background.
```
Comment (carlosapaduarte) [inline _rules/text-contrast-enhanced-09o5cg.md:120]:
```suggestion
This 14pt bold black text has a contrast ratio of 4.6:1 on the gray background.
```

File: _rules/text-contrast-enhanced-09o5cg.md [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+	}
+</style>
+<span id="backgroundSplit">
+	Hello world
+</span>
+```
+
+#### Failed Example 10
+
+The first `p` element has a contrast ratio of 12.6:1. The second `p` element contains an example of the Helvetica font, which has a contrast ratio of 6.4:1. Because this provides information, and not only for aesthetic purposes, this is not considered [purely decorative][].
```
Comment (carlosapaduarte) [inline _rules/text-contrast-enhanced-09o5cg.md:266]:
```suggestion
The first `p` element has a contrast ratio of 12.6:1. The second `p` element, which contains an example of the Helvetica font, has a contrast ratio of 6.4:1. Because this provides information, and not only for aesthetic purposes, this is not considered [purely decorative][].
```

**Review by carlosapaduarte — CHANGES_REQUESTED** [review]

File: _rules/text-contrast-enhanced-09o5cg.md (hunk 1 above)
Comment (Jym77) [reply _rules/text-contrast-enhanced-09o5cg.md:38]: 🤔 Actually wondering if it's worth creating a definition of "semantic widget", the same way we have "semantic link" (in several PRs now).
WDYT?

File: _rules/text-contrast-afw4f7.md [hunk 5]
```diff
@@ -31,7 +31,7 @@ acknowledgments:
 This rule applies to any [visible][] character in a [text node][] that is a [child][] in the [flat tree][] of an HTML element, except if the [text node][] has an [ancestor][] in the [flat tree][] for which one of the following is true:
 
 - **widget**: the ancestor has a [semantic role][] that inherits from `widget`; or
-- **disabled label**: the ancestor is used in the [accessible name][] of a `widget` that is [disabled][]; or
+- **disabled label**: the ancestor is used in the [accessible name][] of an element with a role inheriting from `widget` that is [disabled][]; or
```
Comment (carlosapaduarte) [inline _rules/text-contrast-afw4f7.md:34]: You know it already. Wilco's going to ask you to do this in a different PR ;-)

File: _rules/text-contrast-enhanced-09o5cg.md (hunk 1 above)
Comment (carlosapaduarte) [reply _rules/text-contrast-enhanced-09o5cg.md:38]: I think it is worth it 👍

File: _rules/text-contrast-afw4f7.md (hunk 5 above)
Comment (Jym77) [reply _rules/text-contrast-afw4f7.md:34]: 🙈 I though I did that in #1532

File: pages/glossary/semantic-widget.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+title: Semantic widget
+key: semantic-widget
+unambiguous: true
+objective: true
+input_aspects:
+  - Accessibility tree
+  - DOM tree
+---
+
+A _semantic widget_ is any element whose [semantic role][] is one of the [widget roles][], that is a non-[abstract role][] that inherits from the [`widget`][widget] role.
```
Comment (carlosapaduarte) [inline pages/glossary/semantic-widget.md:11]: I'm not sure we need the second part of the sentence (i.e. after the comma). I don't see the added value in explaining what are "widget roles".
What I would consider adding is a clarification that both "standalone user interface widgets" and "composite user interface widgets" are covered by the definition.

**Review by carlosapaduarte — COMMENTED** [review]
No change requests this time, just something to discuss.

File: pages/glossary/semantic-widget.md (hunk 6 above)
Comment (Jym77) [reply pages/glossary/semantic-widget.md]: Wasn't sure either and went through several iterations. The initial train of thought was:
* Copy semantic link => "role of widget or inheriting from widget".
* But `widget` shouldn't be used as role, so remove it.
* But it looks weird that `widget` is not a "semantic widget", this need clarification.
* Oh wait, there's actually a list already.

🤷 

Maybe it is better now 😁

**Review by carlosapaduarte — APPROVED** [review]

**Review by WilcoFiers — APPROVED** [review]

**Review by nitedog — APPROVED** [review]

**Comment by Jym77** [comment]
Final Call ends on February 1st.

**Comment by Jym77** [comment]
Final Call has ended. Merging.
